src/include/binder/bind_node_visitor.h

```diff
@@ -37,54 +38,80 @@ class BindNodeVisitor : public SqlNodeVisitor {
+  void Visit(common::ManagedPointer<parser::AnalyzeStatement> node,
+             common::ManagedPointer<BinderSherpa> sherpa) override;
+  void Visit(common::ManagedPointer<parser::CaseExpression> expr, common::ManagedPointer<BinderSherpa> sherpa) override;
+  void Visit(common::ManagedPointer<parser::SubqueryExpression> expr,
+             common::ManagedPointer<BinderSherpa> sherpa) override;
+  void Visit(common::ManagedPointer<parser::ConstantValueExpression> expr,
+             common::ManagedPointer<BinderSherpa> sherpa) override;
+  void Visit(common::ManagedPointer<parser::ColumnValueExpression> expr,
+             common::ManagedPointer<BinderSherpa> sherpa) override;
+  void Visit(common::ManagedPointer<parser::StarExpression> expr, common::ManagedPointer<BinderSherpa> sherpa) override;
   // TODO(Ling): implement this after we add support for function expression
-  void Visit(parser::FunctionExpression *expr, parser::ParseResult *parse_result) override;
-  void Visit(parser::OperatorExpression *expr, parser::ParseResult *parse_result) override;
-  void Visit(parser::AggregateExpression *expr, parser::ParseResult *parse_result) override;
-  void Visit(parser::TypeCastExpression *expr, parser::ParseResult *parse_result) override;
+  void Visit(common::ManagedPointer<parser::FunctionExpression> expr,
+             common::ManagedPointer<BinderSherpa> sherpa) override;
+  void Visit(common::ManagedPointer<parser::OperatorExpression> expr,
+             common::ManagedPointer<BinderSherpa> sherpa) override;
+  void Visit(common::ManagedPointer<parser::AggregateExpression> expr,
+             common::ManagedPointer<BinderSherpa> sherpa) override;
+  void Visit(common::ManagedPointer<parser::TypeCastExpression> expr,
+             common::ManagedPointer<BinderSherpa> sherpa) override;
 
  private:
   /** Current context of the query or subquery */
-  BinderContext *context_ = nullptr;
+  common::ManagedPointer<BinderContext> context_ = nullptr;
   /** Catalog accessor */
-  common::ManagedPointer<catalog::CatalogAccessor> catalog_accessor_;
-  /** Default database name of the query. Default to current database reside in */
-  std::string default_database_name_;
+  const common::ManagedPointer<catalog::CatalogAccessor> catalog_accessor_;
+  const catalog::db_oid_t db_oid_;
+
+  void ValidateDatabaseName(const common::ManagedPointer<parser::TableRef> node,
+                            const common::ManagedPointer<BinderSherpa> sherpa) {
+    if (node->table_info_ == nullptr) node->table_info_ = std::make_unique<parser::TableInfo>();
+    if (!(node->GetDatabaseName().empty())) {
+      const auto db_oid = catalog_accessor_->GetDatabaseOid(node->GetDatabaseName());
+      if (db_oid == catalog::INVALID_DATABASE_OID) throw BINDER_EXCEPTION("Database does not exist");
+      if (db_oid != db_oid_) throw BINDER_EXCEPTION("Not connected to specified database");
+    }
```

Comment:
the second if can be put in a separate function. It is used multiple times throughout the code.

Reply:
This is a helper function in the header file / a separate function already. I'm not sure what you mean, do you mean a helper function for the helper function?

Reply:
sorry for not being clear. I mean this part
`        if (!(node->GetDatabaseName().empty())) {
        const auto db_oid = catalog_accessor_->GetDatabaseOid(node->GetDatabaseName());
        if (db_oid == catalog::INVALID_DATABASE_OID) throw BINDER_EXCEPTION("Database does not exist");
        if (db_oid != db_oid_) throw BINDER_EXCEPTION("Not connected to specified database");
      }`
this part got duplicated several times in the bind node visitor especially CreateStatement